File: codeforces/835/G.cpp
#include <iostream>
#include <vector>
#include <set>
using namespace std;

vector<vector<pair<int, int>>> tree;
set<int> pathA;

int n, a, b;

void dfs(int cur, int parent, int x)
{
    if (cur == b)
    {
        return;
    }
    pathA.insert(x);
    for (auto &neighbor : tree[cur])
    {
        int neigh = neighbor.first;
        int w = neighbor.second;
        if (neigh == parent)
        {
            continue;
        }
        dfs(neigh, cur, x ^ w);
    }
}

bool dfsB(int cur, int parent, int x)
{
    if (cur != b && pathA.count(x))
    {
        return true;
    }
    for (auto &neighbor : tree[cur])
    {
        int neigh = neighbor.first;
        int w = neighbor.second;
        if (neigh == parent)
        {
            continue;
        }
        if (dfsB(neigh, cur, x ^ w))
        {
            return true;
        }
    }
    return false;
}

int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        cin >> n >> a >> b;
        tree.assign(n + 1, vector<pair<int, int>>());
        for (int i = 0; i < n - 1; ++i)
        {
            int u, v, w;
            cin >> u >> v >> w;
            tree[u].emplace_back(v, w);
            tree[v].emplace_back(u, w);
        }

        pathA.clear();
        dfs(a, -1, 0);
        if (dfsB(b, -1, 0))
        {
            cout << "YES" << endl;
        }
        else
        {
            cout << "NO" << endl;
        }
    }
    return 0;
}

File: codeforces/835/G.py
# traverse to all node from A, log XOR of path for every node -> store in set
# do the same from B
# if intersection set A and B not empty, print YES

def dfs(cur, parent, x):
    if cur == b:
        return
    pathA.add(x)
    for neigh, w in tree[cur]:
        if neigh == parent:
            continue
        dfs(neigh, cur, x^w)

def dfsB(cur, parent, x):
    if cur != b and x in pathA:
        return True
    for neigh, w in tree[cur]:
        if neigh == parent:
            continue
        if dfsB(neigh, cur, x ^ w):
            return True
    return False

t = int(input())
for _ in range(t):
    n, a, b = map(int, input().split())
    tree = [[] for i in range(n + 1)]
    for i in range(n - 1):
        u, v, w = map(int, input().split())
        tree[u].append((v, w))
        tree[v].append((u, w))

    pathA = set()
    dfs(a, -1, 0)
    if dfsB(b, -1, 0):
        print("YES")
    else:
        print("NO")
